Mathlib/Algebra/Module/Submodule/LinearMap.lean

```diff
@@ -0,0 +1,476 @@
+#align linear_map.smul_right LinearMap.smulRight
+
+@[simp]
+theorem coe_smulRight (f : M₁ →ₗ[R] S) (x : M) : (smulRight f x : M₁ → M) = fun c => f c • x :=
+  rfl
+#align linear_map.coe_smul_right LinearMap.coe_smulRight
+
+theorem smulRight_apply (f : M₁ →ₗ[R] S) (x : M) (c : M₁) : smulRight f x c = f c • x :=
+  rfl
+#align linear_map.smul_right_apply LinearMap.smulRight_apply
+
+end SMulRight
+
+instance [Nontrivial M] : Nontrivial (Module.End R M) := by
+  obtain ⟨m, ne⟩ := exists_ne (0 : M)
+  exact nontrivial_of_ne 1 0 fun p => ne (LinearMap.congr_fun p m)
+
+@[simp, norm_cast]
+theorem coeFn_sum {ι : Type*} (t : Finset ι) (f : ι → M →ₛₗ[σ₁₂] M₂) :
+    ⇑(∑ i in t, f i) = ∑ i in t, (f i : M → M₂) :=
+  _root_.map_sum
+    (show AddMonoidHom (M →ₛₗ[σ₁₂] M₂) (M → M₂)
+      from { toFun := FunLike.coe,
+             map_zero' := rfl
+             map_add' := fun _ _ => rfl }) _ _
+#align linear_map.coe_fn_sum LinearMap.coeFn_sum
+
+theorem submodule_pow_eq_zero_of_pow_eq_zero {N : Submodule R M} {g : Module.End R N}
+    {G : Module.End R M} (h : G.comp N.subtype = N.subtype.comp g) {k : ℕ} (hG : G ^ k = 0) :
+    g ^ k = 0 := by
+  ext m
+  have hg : N.subtype.comp (g ^ k) m = 0 := by
+    rw [← commute_pow_left_of_commute h, hG, zero_comp, zero_apply]
+  simpa using hg
+#align linear_map.submodule_pow_eq_zero_of_pow_eq_zero LinearMap.submodule_pow_eq_zero_of_pow_eq_zero
+
+section
+
+variable {f' : M →ₗ[R] M}
+
+theorem pow_apply_mem_of_forall_mem {p : Submodule R M} (n : ℕ) (h : ∀ x ∈ p, f' x ∈ p) (x : M)
+    (hx : x ∈ p) : (f' ^ n) x ∈ p := by
+  induction' n with n ih generalizing x
+  · simpa
+  · simpa only [iterate_succ, coe_comp, Function.comp_apply, restrict_apply] using ih _ (h _ hx)
+#align linear_map.pow_apply_mem_of_forall_mem LinearMap.pow_apply_mem_of_forall_mem
+
+theorem pow_restrict {p : Submodule R M} (n : ℕ) (h : ∀ x ∈ p, f' x ∈ p)
+    (h' := pow_apply_mem_of_forall_mem n h) :
+    (f'.restrict h) ^ n = (HPow.hPow f' n).restrict h' := by
+  ext x
+  have : Semiconj (↑) (f'.restrict h) f' := fun _ ↦ restrict_coe_apply _ _ _
+  simp [coe_pow, this.iterate_right _ _]
+#align linear_map.pow_restrict LinearMap.pow_restrict
+
+end
+
+end AddCommMonoid
+
+section Module
+
+variable [Semiring R] [Semiring S] [AddCommMonoid M] [AddCommMonoid M₂] [AddCommMonoid M₃]
+  [Module R M] [Module R M₂] [Module R M₃] [Module S M₂] [Module S M₃] [SMulCommClass R S M₂]
+  [SMulCommClass R S M₃] (f : M →ₗ[R] M₂)
+
+
+variable (S)
+
+/-- Applying a linear map at `v : M`, seen as `S`-linear map from `M →ₗ[R] M₂` to `M₂`.
+
+ See `LinearMap.applyₗ` for a version where `S = R`. -/
+@[simps]
+def applyₗ' : M →+ (M →ₗ[R] M₂) →ₗ[S] M₂ where
+  toFun v :=
+    { toFun := fun f => f v
+      map_add' := fun f g => f.add_apply g v
+      map_smul' := fun x f => f.smul_apply x v }
+  map_zero' := LinearMap.ext fun f => f.map_zero
+  map_add' _ _ := LinearMap.ext fun f => f.map_add _ _
+#align linear_map.applyₗ' LinearMap.applyₗ'
```

Comment:
This doesn't seem to involve submodules? It can probably go all the way to the basic linear map file. (This seems to hold for all changes in `c809cea5477043491f1363cc7ecfd43e7f7e8ed3`.)

Reply:
`LinearMap.ringLmapEquivSelf` needed linear equivalences, so I moved it to `Mathlib/Algebra/Module/Equiv.lean`. 

`LinearMap.domRestrict'` does need submodules, so I left it in `Mathlib/Algebra/Module/Submodule/LinearMap.lean`.

Moved everything else.